indra/llinventory/llinventory.cpp

```diff
@@ -1050,6 +1052,159 @@ bool LLInventoryItem::fromLLSD(const LLSD& sd, bool is_new)
+    end = sd.endMap();
+    for (i = sd.beginMap(); i != end; ++i)
+    {
+        if (i->first == INV_ITEM_ID_LABEL)
+        {
+            mUUID = i->second;
+        }
+
+        if (i->first == INV_PARENT_ID_LABEL)
+        {
+            mParentUUID = i->second;
+        }
+
+        if (i->first == INV_THUMBNAIL_LABEL)
+        {
+            const LLSD &thumbnail_map = i->second;
+            const std::string w = INV_ASSET_ID_LABEL;
+            if (thumbnail_map.has(w))
+            {
+                mThumbnailUUID = thumbnail_map[w];
+            }
+            /* Example:
+                <key> asset_id </key>
+                <uuid> acc0ec86 - 17f2 - 4b92 - ab41 - 6718b1f755f7 </uuid>
+                <key> perms </key>
+                <integer> 8 </integer>
+                <key>service</key>
+                <integer> 3 </integer>
+                <key>version</key>
+                <integer> 1 </key>
+            */
+        }
+
+        if (i->first == INV_THUMBNAIL_ID_LABEL)
+        {
+            mThumbnailUUID = i->second.asUUID();
+        }
+
+        if (i->first == INV_PERMISSIONS_LABEL)
+        {
+            mPermissions = ll_permissions_from_sd(i->second);
+        }
+
+        if (i->first == INV_SALE_INFO_LABEL)
+        {
+            // Sale info used to contain next owner perm. It is now in
+            // the permissions. Thus, we read that out, and fix legacy
+            // objects. It's possible this op would fail, but it
+            // should pick up the vast majority of the tasks.
+            BOOL has_perm_mask = FALSE;
+            U32  perm_mask     = 0;
+            if (!mSaleInfo.fromLLSD(i->second, has_perm_mask, perm_mask))
+            {
+                goto fail;
+            }
+            if (has_perm_mask)
+            {
+                if (perm_mask == PERM_NONE)
+                {
+                    perm_mask = mPermissions.getMaskOwner();
+                }
+                // fair use fix.
+                if (!(perm_mask & PERM_COPY))
+                {
+                    perm_mask |= PERM_TRANSFER;
+                }
+                mPermissions.setMaskNext(perm_mask);
+            }
+        }
+
+        if (i->first == INV_SHADOW_ID_LABEL)
+        {
+            mAssetUUID = i->second;
+            LLXORCipher cipher(MAGIC_ID.mData, UUID_BYTES);
+            cipher.decrypt(mAssetUUID.mData, UUID_BYTES);
+        }
+
+        if (i->first == INV_ASSET_ID_LABEL)
+        {
+            mAssetUUID = i->second;
+        }
+
+        if (i->first == INV_LINKED_ID_LABEL)
+        {
+            mAssetUUID = i->second;
+        }
+
+        if (i->first == INV_ASSET_TYPE_LABEL)
+        {
+            LLSD const &label = i->second;
+            if (label.isString())
+            {
+                mType = LLAssetType::lookup(label.asString().c_str());
+            }
+            else if (label.isInteger())
+            {
+                S8 type = (U8) label.asInteger();
```

Comment:
Why not cast to S8 directly?

Reply:
Hmm.  unfortunately, I don't know why it was originally implemented this way.  I copied this code directly from the above block that is now in the `#if 0` and tried to disturb it as little as possible while restructuring the LLSD traversal code for this optimization.  I did not dig in to the business logic of these pieces and attempted to preserve them as-is.